core/iwasm/common/wasm_memory.c

```diff
@@ -54,7 +54,7 @@ wasm_memory_init_with_allocator(void *_user_data, void *_malloc_func,
                                 void *_realloc_func, void *_free_func)
 {
     if (_malloc_func && _free_func && _malloc_func != _free_func) {
-        memory_mode = MEMORY_MODE_ALLOCATOR;
+        __atomic_store_n(&memory_mode, MEMORY_MODE_ALLOCATOR, __ATOMIC_SEQ_CST);
```

Comment:
This won't work with e.g. msvc compiler, although I'm not sure if wasi-libc is supposed to work on windows? Anyway, as @wenyongh pointed out, we might not need this change if his PR fixes the problem.